internal/pkg/mining/scheduler.go

```diff
@@ -7,108 +7,127 @@ package mining
-	// we buffer 1 to make sure we do not get blocked when shutting down
+// Start starts mining. It takes in a channel signaling when to skip mining --
+// intended to be wired to the syncer's catchup notifications -- and a context.
+// It returns a channel for reading mining outputs and a waitgroup for teardown.
+// It is the callers responsibility to close the out channel ONLY after waiting
+// on the waitgroup.
+func (s *timingScheduler) Start(miningCtx context.Context, skipCh chan bool) (<-chan Output, *sync.WaitGroup) {
+	var doneWg sync.WaitGroup
 	outCh := make(chan Output, 1)
-	var doneWg sync.WaitGroup    // for internal use
-	var extDoneWg sync.WaitGroup // for external use
 
-	log.Debugf("Scheduler starting main receive loop")
+	// loop mining work
 	doneWg.Add(1)
-
-	s.isStarted = true
 	go func() {
-		defer doneWg.Done()
-		nullBlkCount := uint64(0)
-		var prevBase block.TipSet
-		var prevWon bool
-		for {
-			select {
-			case <-miningCtx.Done():
-				s.isStarted = false
-				return
-			default:
-			}
-			// This is the sleep during which we collect. TODO: maybe this should vary?
-			time.Sleep(s.mineDelay)
-			// Ask for the heaviest tipset.
-			base, _ := s.pollHeadFunc()
-			if !base.Defined() { // Don't try to mine on an unset head.
-				outCh <- NewOutput(nil, errors.New("cannot mine on unset (nil) head"))
-				return
-			}
-			if prevWon && prevBase.Equals(base) {
-				// Skip this round, this likely means that the new head has not propagated yet through the system.
-				// TODO: investigate if there is a better way to handle this situation.
-				continue
-			}
-
-			// Determine how many null blocks we should mine with.
-			nullBlkCount = nextNullBlkCount(nullBlkCount, prevBase, base)
-
-			// Mine synchronously! Ignore all new tipsets.
-			prevWon = s.worker.Mine(miningCtx, base, nullBlkCount, outCh)
-			prevBase = base
-		}
+		s.mineLoop(miningCtx, outCh, skipCh, &doneWg)
+		doneWg.Done()
 	}()
+	s.isStarted = true
 
-	// This tear down goroutine waits for all work to be done before closing
-	// channels.  When this goroutine is complete, external code can
-	// consider the scheduler to be done.
-	extDoneWg.Add(1)
-	go func() {
-		defer extDoneWg.Done()
-		doneWg.Wait()
-		close(outCh)
-	}()
-	return outCh, &extDoneWg
+	return outCh, &doneWg
+}
+
+func (s *timingScheduler) mineLoop(miningCtx context.Context, outCh chan Output, skipCh chan bool, doneWg *sync.WaitGroup) {
+	// mineLoop is the main event loop for the timing scheduler.  It waits for
+	// a new epoch to start, polls the heaviest head, includes the correct number
+	// of null blocks and starts a mining job async.
+	//
+	// If the previous epoch's mining job is not finished it is canceled via the context
+	//
+	// The scheduler will skip mining jobs if the skipping flag is set
+	workContext, workCancel := context.WithCancel(miningCtx) // nolint:staticcheck
+	for {
+		s.waitForEpochStart(miningCtx)
```

Comment:
I wouldn't just flatout wait for an epoch, I would:

1.  track the epoch we started mining at
2. mine 
3. check the epoch at the time we finish 
   3.a.1 if we are still in the same epoch, all good, do the post-mining work

   3.b.1 if we are NOT in the same epoch, we missed the window, and the mining wont work 
   3.b.2 log a warning that we failed to produce a block during the epoch mining window. 

4. wait for next epoch
> // Note: we wait even in the case were we are already on the next window to have a chance at a full window to mine a block`

Reply:
We *are* doing a slightly better version of what you suggest.  We use the goroutine and cancel function to communicate that the time has passed and stop the mining work from the previous epoch which should log the warning.  Might need to fiddle with log statements to get exactly what we want but that stuff can wait after this PR.

The only difference between whats here and whats proposed is that we don't block new mining work on late mining work from a previous epoch which is a better way to handle this by default.  As you say if we end up being late "mining won't work" so it's better to not block tasks that might work on these tasks that we agree won't work.